### PR TITLE
Changed the facilitators selector into a search bar

### PR DESCRIPTION
**Pull request by mohamudm1**
Solves issue #417. 
Changed the facilitators selector in the created events page into a search bar where facilitators can be searched.

### CODE REVIEW

**Comment by BrianRamsay** [comment]
merge conflict to resolve

File: app/logic/events.py [hunk 1]
```diff
@@ -325,16 +323,39 @@ def preprocessEventData(eventData):
...
-        eventData['facilitators'] = list(User.select().join(EventFacilitator).where(EventFacilitator.event == event))
+    if 'facilitators' in eventData:
+        eventData['facilitators'] = getFacilitatorsFromList(eventData['facilitators'])
+    # try:
+    #     if type(eventData) == MultiDict and type(eventData['facilitators']) is not list:
+    #         eventData['facilitators'] = eventData.getlist('facilitators')
+    #     eventData['facilitators'] = [User.get_by_id(f) for f in eventData['facilitators']]
+    # except Exception as e:
+    #     event = eventData.get('id', -1)
+    #     eventData['facilitators'] = list(User.select().join(EventFacilitator).where(EventFacilitator.event == event))
```
Comment (BrianRamsay) [inline app/logic/events.py]: Don't leave a bunch of commented out code.

File: app/logic/events.py [hunk 2]
```diff
@@ -325,16 +323,39 @@ def preprocessEventData(eventData):
...
+    """
+    This function takes in a list of usernames or a string with the usernames of
+        facilitators separated by commas (,) and returns a list of facilitator
+        objects that match the usernames in the list
+    facilitatorList: expected to be a list with facilitator usernames as strings
+    return: list of facilitator objects matching the usernames in facilitatorList
+    """
+    if type(facilitatorList) == str:
+        facilitatorList = facilitatorList.split(',')
+    print(facilitatorList)
```
Comment (BrianRamsay) [inline app/logic/events.py]: remove print

File: app/logic/events.py [hunk 3]
```diff
@@ -325,16 +323,39 @@ def preprocessEventData(eventData):
...
+    #     eventData['facilitators'] = list(User.select().join(EventFacilitator).where(EventFacilitator.event == event))
 
     return eventData
 
+def getFacilitatorsFromList(facilitatorList):
+    """
+    This function takes in a list of usernames or a string with the usernames of
+        facilitators separated by commas (,) and returns a list of facilitator
+        objects that match the usernames in the list
+    facilitatorList: expected to be a list with facilitator usernames as strings
```
Comment (BrianRamsay) [inline app/logic/events.py:336]: Your function also handles a comma separated list of usernames

Comment (BrianRamsay) [reply app/logic/events.py:336]: Make sure your comment is consistent

File: app/logic/events.py [hunk 4]
```diff
@@ -325,16 +323,39 @@ def preprocessEventData(eventData):
...
+    return: list of facilitator objects matching the usernames in facilitatorList
+    """
+    if type(facilitatorList) == str:
+        facilitatorList = facilitatorList.split(',')
+    print(facilitatorList)
+    finalFacilitatorList = []
+    for i in facilitatorList:
```
Comment (BrianRamsay) [inline app/logic/events.py]: Use a descriptive name instead of `i`. `i` is only for simple incrementing numbers

File: app/logic/events.py [hunk 5]
```diff
@@ -325,16 +323,39 @@ def preprocessEventData(eventData):
...
+        if i:
+            facilitatorToAdd = User.select().where(User.username == i).get()
```
Comment (BrianRamsay) [inline app/logic/events.py]: This can just be `User.get_by_id(username)`

File: app/logic/events.py (hunk 4 above)
Comment (BrianRamsay) [reply app/logic/events.py]: This is a good case for a list comprehension instead of a for loop

File: app/static/js/createEvents.js [hunk 6]
```diff
@@ -127,4 +129,39 @@ $(document).ready(function() {
...
+        newRow.attr("id", username);
+        newRow.prop("hidden", false);
+        lastRow.after(newRow);
+        $("#hiddenFacilitatorArray").attr("value", facilitatorArray);
+    }
+  }
+
+  $("#eventFacilitator").on('input', function() {
+    // To retrieve specific columns into a dict, create a [] list and put columns inside
+    searchUser("eventFacilitator", callback, true, null, ["firstName", "lastName", "username"]);
```
Comment (BrianRamsay) [inline app/static/js/createEvents.js]: That last parameter isn't necessary anymore. Also, don't you need a category here?

File: app/static/js/searchStudent.js [hunk 7]
```diff
@@ -4,5 +4,5 @@ function callback(selected) {
 }
 
 $("#searchStudentsInput").on("input", function() {
-  searchUser("searchStudentsInput", callback);
+  searchUser("searchStudentsInput", callback, undefined, undefined, undefined,"students");
```
Comment (BrianRamsay) [inline app/static/js/searchStudent.js]: it searches students by default

File: app/templates/admin/createEvent.html [hunk 8]
```diff
@@ -13,6 +13,8 @@
 {% endblock %}
 
 {% block app_content %}
+{% set isAdmin = g.current_user.isCeltsAdmin or isProgramManager %}
+{% set isEditable = isAdmin %}
```
Comment (BrianRamsay) [inline app/templates/admin/createEvent.html]: I think this is irrelevant now that there is a separate view template

File: app/templates/admin/createEvent.html [hunk 9]
```diff
@@ -152,17 +154,58 @@ <h1 id="title">{{page_title}}</h1>
...
+            <input
+              id="eventFacilitator"
+              class="form-control"
+              type="search"
+              placeholder="Search facilitator"
+              value="" />
+            <table
+              class="table mb-3"
+              id="facilitatorTable"
+              name="facilitatorTable">
```
Comment (BrianRamsay) [inline app/templates/admin/createEvent.html]: if the attributes fit on one row, leave them on one row

File: app/templates/admin/createEvent.html [hunk 10]
```diff
@@ -152,17 +154,58 @@ <h1 id="title">{{page_title}}</h1>
...
+                {% endfor %}
+              </tbody>
+            </table>
+            {% else %}
+              {% if not eventData.facilitators|count %}
+                <p>None</p>
+                {%endif%}
+                {% for f in eventData.facilitators %}
+                <p>{{f.firstName}} {{f.lastName}}</p>
+                {% endfor %}
```
Comment (BrianRamsay) [inline app/templates/admin/createEvent.html]: This bit should go in the view template

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/templates/admin/createEvent.html (hunk 10 above)
Comment (Lyisus) [reply app/templates/admin/createEvent.html]: This part should be kept here for when the event edited. If there are facilitators, the person editing the event should be able to see them. However, this currently does not work with the new search bar, but I think Mostaphe created an issue for it to keep this PR from getting overly large.

File: app/logic/events.py [hunk 11]
```diff
@@ -325,16 +323,27 @@ def preprocessEventData(eventData):
...
+    This function takes in a list of usernames or a string with the usernames of
+        facilitators separated by commas (,) and returns a list of facilitator
+        objects that match the usernames in the list
+    facilitatorList: expected to be a list with facilitator usernames as strings
+                        or a string of usernames separated by commas (,)
+    return: list of facilitator objects matching the usernames in facilitatorList
+    """
+    if facilitatorList and type(facilitatorList) == str:
+        facilitatorList = facilitatorList.split(',')
+    finalFacilitatorList = []
```
Comment (BrianRamsay) [inline app/logic/events.py]: This isn't necessary anymore

File: app/logic/events.py [hunk 12]
```diff
@@ -325,16 +323,27 @@ def preprocessEventData(eventData):
...
+    # for each facilitator in facilitatorList, if it is not a blank string of null
+    # add a user object matching said facilitator to finalFacilitatorList
+    finalFacilitatorList = [User.get_by_id(facilitator) for facilitator in facilitatorList if facilitator]
```
Comment (BrianRamsay) [inline app/logic/events.py:344]: Is there a case where we need that `if facilitator`? We probably want to have an error if there is weird data being passed in

File: app/templates/admin/createEvent.html (hunk 10 above)
Comment (BrianRamsay) [reply app/templates/admin/createEvent.html]: This page is always editable. That's why there aren't any more isEditable references

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/templates/admin/createEvent.html [hunk 13]
```diff
@@ -152,17 +154,40 @@ <h1 id="title">{{page_title}}</h1>
...
+              </thead>
+              <tbody>
+                <tr hidden>
+                  <td>
+                    <p class="facilitatorName" required></p>
+                  </td>
+                  <td><button id="remove" type="button" class="btn btn-danger">Remove</button></td>
+
+                </tr>
+                {% for instructor in courseInstructor %}
```
Comment (BrianRamsay) [inline app/templates/admin/createEvent.html:177]: Facilitators aren't showing up on the create event page. I can see it was added by looking at View, but it's not there on Edit

File: app/templates/admin/createEvent.html [hunk 14]
```diff
@@ -152,17 +154,40 @@ <h1 id="title">{{page_title}}</h1>
...
+              id="eventFacilitator"
+              class="form-control"
+              type="search"
+              placeholder="Search facilitator"
+              value="" />
+            <table class="table mb-3" id="facilitatorTable" name="facilitatorTable">
+              <thead>
+                <tr>
+                  <th scope="col">Facilitator</th>
+                  <th scope="col">Remove</th>
```
Comment (BrianRamsay) [inline app/templates/admin/createEvent.html]: Let's remove the header text above the buttons

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/templates/admin/createEvent.html (hunk 13 above)
Comment (Lyisus) [reply app/templates/admin/createEvent.html:177]: We made that a separate issue to keep this PR smaller. We'll get to that one once this one is merged.

File: app/static/js/createEvents.js [hunk 15]
```diff
@@ -127,4 +129,39 @@ $(document).ready(function() {
       $("#checkIsRequired").prop('disabled', false);
     }
   });
+
+  var facilitatorArray = []
+  function callback(selectedFacilitator) {
+    // JSON.parse is required to de-stringify the search results into a dictionary.
```
Comment (BrianRamsay) [inline app/static/js/createEvents.js:135]: This comment isn't true anymore

**Review by BrianRamsay — APPROVED** [review]

**Pull request merged**